hazelcast/src/main/java/com/hazelcast/kubernetes/KubernetesRestClient.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.kubernetes;
+
+import com.hazelcast.internal.nio.IOUtil;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.spi.utils.AbstractRestClient;
+
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSocketFactory;
+import javax.net.ssl.TrustManagerFactory;
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import java.security.cert.CertificateException;
+import java.security.cert.CertificateFactory;
+import java.util.Collection;
+
+/**
+ * Utility class for making REST calls.
+ */
+final class KubernetesRestClient extends AbstractRestClient<KubernetesRestClient> {
```

Comment:
I wonder if we need to have this class at all? I think all the code is generic, in a sense that could be used as well outside of the Kubernetes context. What if we put everything in the `RestClient` class and in the `buildHttpConnection` we check add the following check? Wouldn't it work?
```
        if (connection instanceof HttpsURLConnection && caCertificate != null) {
            ((HttpsURLConnection) connection).setSSLSocketFactory(buildSslSocketFactory());
        }
```

The thing is that from the class structure someone could think that Kubernetes requires some special Rest client interaction, while in fact Kubernetes is just a standard Rest Client, it just needs mutual TLS (which is not used in at the moment in AWS/GCP/Azure).

Reply:
Yes, but the `buildSslSocketFactory` throws `KubernetesClientException`, and changing it to throw `RestClientExcetion` will break backward compatibility. If we are ok with this change, then we can move to one class and unify it